## NumPy

NumPy (short for **Numerical Python**) is a powerful open-source Python library used for numerical computing. It provides tools for creating and manipulating arrays and matrices, which are collections of values organized in rows and columns, similar to spreadsheets or tables. It is widely used in data analysis, machine learning, scientific computing, and engineering.

**Why Use NumPy?**
1. Efficient Numerical Computations:
    - NumPy arrays (ndarrays) are faster and more memory-efficient than Python lists because they are implemented in C and optimized for numerical operations.
1. Array Operations:
    - Supports element-wise operations, such as addition, subtraction, multiplication, and division, directly on arrays.
1. Mathematical Functions:
    - Provides a wide range of mathematical functions for operations like linear algebra, Fourier transforms, and statistical computations.
1. Multidimensional Arrays:
    - Allows for creating and manipulating arrays of any dimension (1D, 2D, 3D, etc.), enabling complex computations on large datasets.
1. Broadcasting:
    - Enables operations between arrays of different shapes, simplifying computations without the need for explicit loops.
1. Integration with Other Libraries:
    - Serves as a foundation for many other Python libraries, including Pandas, Scikit-learn, TensorFlow, and Matplotlib.

**When Should You Use NumPy?**
- When working with large datasets that require efficient computations.
- For tasks involving matrix operations, linear algebra, or scientific calculations.
- As a base for data manipulation in libraries like Pandas or for preprocessing in machine learning pipelines.

In [1]:
# Import the numpy package as np, so that you can refer to numpy with np.
import numpy as np

In [2]:
np_arr = np.array([1,2,3,4,5])
print(np_arr)
print(np_arr.dtype)

# Check the type of array
type(np_arr)

[1 2 3 4 5]
int32


numpy.ndarray

### Comparison of List & NumPy

In [3]:
# Let create a list
arr1 = [3,2,6,4,1]

print(arr1)
print(type(arr1))

[3, 2, 6, 4, 1]
<class 'list'>


In [4]:
arr2 = [4,3,2,3,6]

# Add list
print(arr1 + arr2)

[3, 2, 6, 4, 1, 4, 3, 2, 3, 6]


In [5]:
# multiple list
print(arr1 * 2)

[3, 2, 6, 4, 1, 3, 2, 6, 4, 1]


In [6]:
np_arr1 = np.array(arr1)
np_arr2 = np.array(arr2)
print(np_arr1)
print(np_arr2)

[3 2 6 4 1]
[4 3 2 3 6]


In [7]:
# Add to array
print(np_arr1 + np_arr2)

[7 5 8 7 7]


In [8]:
# Multiple array
print(np_arr1)
print(np_arr1 * 2)

[3 2 6 4 1]
[ 6  4 12  8  2]


### Dataype in NumPy

In [9]:
dt = np.dtype([('age',np.int8)])
dt

dtype([('age', 'i1')])

In [10]:
type((10,))

tuple

In [11]:
a = np.array([(10,),(20,),(30,)], dtype = dt) 
print(a)

[(10,) (20,) (30,)]


In [12]:
dt = np.dtype([('name','S10'),('age',np.int8)])
dt

dtype([('name', 'S10'), ('age', 'i1')])

In [13]:
b = np.array([('abc',1),('xyz',2)], dtype=dt)
print(b)

[(b'abc', 1) (b'xyz', 2)]


In [14]:
arr = np.array(['apple','banana','orange'])
print(arr.dtype)

<U6


### 2D NumPy Arrays

In [15]:
arr = np.array([[1,2,3],[4,5,6]])
print(arr)
print(arr.dtype)

[[1 2 3]
 [4 5 6]]
int32


ndarray.shape

In [16]:
# shape array attribute returns a tuple consisting of array dimensions.
# ndarray.shape

print(arr.shape)

(2, 3)


In [17]:
# resize the array

arr.shape = (3,2)
print(arr)

[[1 2]
 [3 4]
 [5 6]]


ndarray.ndim

In [18]:
# This array attribute returns the number of array dimensions.
arr = np.arange(12) 
print(f"Arry dimension:",arr.ndim)
print("Array element: ", arr)
print("Array Size:", arr.size)

Arry dimension: 1
Array element:  [ 0  1  2  3  4  5  6  7  8  9 10 11]
Array Size: 12


In [19]:
# now reshape it 
b = arr.reshape(3,2,2) 
print(f"Arry dimension:",b.ndim)
print("Array element: ", b)
print("Array Size:", b.size)
# b is having three dimensions

Arry dimension: 3
Array element:  [[[ 0  1]
  [ 2  3]]

 [[ 4  5]
  [ 6  7]]

 [[ 8  9]
  [10 11]]]
Array Size: 12


numpy.itemsize

In [20]:
# This attribute will give the size of element in bytes.
print(arr.itemsize)
print(b.itemsize)

4
4


### Array Creation Routines

numpy.empty

In [21]:
# It creates an uninitialized array of specified shape and dtype.
arr = np.empty((3,2), dtype = float)
arr

array([[6.23042070e-307, 4.67296746e-307],
       [1.69121096e-306, 3.33772113e-307],
       [2.67018913e-306, 1.21378159e-311]])

numpy.zero

In [22]:
arr = np.zeros(5) 
arr

array([0., 0., 0., 0., 0.])

In [23]:
arr = np.zeros((2,3,3), dtype=int) 
arr

array([[[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]],

       [[0, 0, 0],
        [0, 0, 0],
        [0, 0, 0]]])

numpy.ones

In [24]:
#  new array of specified size and type, filled with ones.
arr = np.ones(5) 
arr

array([1., 1., 1., 1., 1.])

In [25]:
arr = np.ones((2,2), dtype= np.int32) 
arr

array([[1, 1],
       [1, 1]])

In [26]:
arr = np.identity(3)
arr

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [27]:
arr.flatten()

array([1., 0., 0., 0., 1., 0., 0., 0., 1.])

### Subsetting

Indexing & Slicing

In [28]:
arr = np.arange(10) 
print("Array:", arr)
s = slice(2,8,2) 
arr[s]

Array: [0 1 2 3 4 5 6 7 8 9]


array([2, 4, 6])

In [29]:
arr = np.arange(10) 
slice_arr = arr[2:8:2]
slice_arr

array([2, 4, 6])

In [30]:
print(arr[5])
print(arr)
print(arr[0:4])

5
[0 1 2 3 4 5 6 7 8 9]
[0 1 2 3]


In [31]:
arr_2d = np.array(
    [
        [1.2, 1.3, 1.4, 1.5, 1.6],
        [5.4, 5.5, 5.6, 5.7, 5.8],
        [9.6, 9.7, 9.8, 9.9, 9.10]
    ]
)

arr_2d

array([[1.2, 1.3, 1.4, 1.5, 1.6],
       [5.4, 5.5, 5.6, 5.7, 5.8],
       [9.6, 9.7, 9.8, 9.9, 9.1]])

In [32]:
print(arr_2d[0][2])
print(arr_2d[0,2])

1.4
1.4


In [33]:
print("First row:",arr_2d[0])
print("First row, all column", arr_2d[0,...])
print("First & Second row:",arr_2d[:2])

First row: [1.2 1.3 1.4 1.5 1.6]
First row, all column [1.2 1.3 1.4 1.5 1.6]
First & Second row: [[1.2 1.3 1.4 1.5 1.6]
 [5.4 5.5 5.6 5.7 5.8]]


In [34]:
arr_2d

array([[1.2, 1.3, 1.4, 1.5, 1.6],
       [5.4, 5.5, 5.6, 5.7, 5.8],
       [9.6, 9.7, 9.8, 9.9, 9.1]])

In [35]:
# Transpos array
arr_2d.T

array([[1.2, 5.4, 9.6],
       [1.3, 5.5, 9.7],
       [1.4, 5.6, 9.8],
       [1.5, 5.7, 9.9],
       [1.6, 5.8, 9.1]])

In [36]:
print(arr_2d.T[0])
print(arr_2d[...,0])
print(arr_2d[:,:1].T[0])

[1.2 5.4 9.6]
[1.2 5.4 9.6]
[1.2 5.4 9.6]


Integer Indexing

In [37]:
arr = np.array(
    [
        [1, 2, 3], 
        [4, 5, 6], 
        [7, 8, 9]
    ]
) 
print(arr)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [38]:
narr = arr[[0,1,2], [0,1,1]] 
narr

array([1, 5, 8])

In [39]:
r = np.array([[0,0],[2,2]])
c = np.array([[0,2],[0,2]]) 

corner_arr = arr[r,c] 

print("Array: \n", arr)
print("Corner Array: \n", corner_arr)

Array: 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Corner Array: 
 [[1 3]
 [7 9]]


In [40]:
sub_arr = arr[0:2, 1:3]
sub_arr

array([[2, 3],
       [5, 6]])

Boolean Array Indexing

In [41]:
arr = np.arange(2,21,2)
arr

array([ 2,  4,  6,  8, 10, 12, 14, 16, 18, 20])

In [42]:
arr = arr.reshape(2,5)
arr

array([[ 2,  4,  6,  8, 10],
       [12, 14, 16, 18, 20]])

In [43]:
for x in np.nditer(arr):
   print(x, end=' '),

2 4 6 8 10 12 14 16 18 20 

In [44]:
arr = arr.T
arr

array([[ 2, 12],
       [ 4, 14],
       [ 6, 16],
       [ 8, 18],
       [10, 20]])

In [45]:
for x in np.nditer(arr):
   print(x, end=' '),

2 4 6 8 10 12 14 16 18 20 

In the context of ```np.nditer```, the ```order``` parameter specifies the iteration order over the elements of the array. It controls how the iterator traverses the array elements in memory. The ```order``` parameter can take the following values:

**'C' (Row-Major Order)**
- Iterates through the array in row-major order, which means it accesses the elements in the same order as they are stored in memory for a C-contiguous array.
- This corresponds to the traditional way of traversing a 2D array row by row (left to right, then top to bottom).

In [46]:
for x in np.nditer(arr, order='C'):
    print(x, end=' ')

2 12 4 14 6 16 8 18 10 20 

**'F' (Column-Major Order)**
Iterates through the array in column-major order, meaning it accesses the elements in the same order as they are stored in memory for a Fortran-contiguous array.
This corresponds to traversing the array column by column (top to bottom, then left to right).

In [47]:
for x in np.nditer(arr, order='F'):
   print(x, end=' '),

2 4 6 8 10 12 14 16 18 20 

In [48]:
arr = np.arange(2,21,2)
arr = arr.reshape(2,5)

print(arr, end='\n\n')

# one-dimensional arrays corresponding to each column is traversed by the iterator.

for x in np.nditer(arr, flags = ['external_loop'], order = 'F'):
    print(x)

[[ 2  4  6  8 10]
 [12 14 16 18 20]]

[ 2 12]
[ 4 14]
[ 6 16]
[ 8 18]
[10 20]


numpy.stack

In [49]:
arr1 = np.array([[1,2],[3,4]]) 
arr2 = np.array([[5,6],[7,8]]) 

print('arr1:\n', arr1, end='\n\n')
print('arr2:\n', arr2, end='\n\n')

# Stack the two arrays along axis 0
stack_arr = np.stack((arr1,arr2),0) 
print('Stacked array along axis 0:\n', stack_arr)

# Stack the two arrays along axis 1

stack_arr = np.stack((arr1,arr2),1) 
print('\n\nStacked array along axis 1:\n', stack_arr)

arr1:
 [[1 2]
 [3 4]]

arr2:
 [[5 6]
 [7 8]]

Stacked array along axis 0:
 [[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


Stacked array along axis 1:
 [[[1 2]
  [5 6]]

 [[3 4]
  [7 8]]]


Spliting array

In [50]:
arr = np.arange(16) 

print(arr)

# Split the array in 4 equal-sized subarrays
arr_size = np.split(arr,4) 
print(arr_size)
print('\n')  

# Split the array at positions indicated in 1-D array
arr_pos = np.split(arr,[3,8,12])
print(arr_pos)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]
[array([0, 1, 2, 3]), array([4, 5, 6, 7]), array([ 8,  9, 10, 11]), array([12, 13, 14, 15])]


[array([0, 1, 2]), array([3, 4, 5, 6, 7]), array([ 8,  9, 10, 11]), array([12, 13, 14, 15])]


### Broadcasting Iteration

In [51]:
arr = np.arange(2,31, 2)
arr = arr.reshape(3,5)
arr

array([[ 2,  4,  6,  8, 10],
       [12, 14, 16, 18, 20],
       [22, 24, 26, 28, 30]])

In [52]:
b_arr = np.array([1,2,3,4,5])
b_arr

array([1, 2, 3, 4, 5])

In [53]:
for x,y in np.nditer([arr,b_arr], order='F'): # Column-Major
   print(x*y)

2
12
22
8
28
48
18
48
78
32
72
112
50
100
150


### Mathematical functions in Numpy

In [54]:
np.tan(np.pi/4)

0.9999999999999999

In [55]:
np.sin(np.pi/4)

0.7071067811865476

In [56]:
np.sinh(0)

0.0

In [57]:
np.cosh(0)

1.0

Rounding Functions

In [58]:
f_array = np.array ([0.1485, 1.1726, 2.99 , -3.2])

In NumPy, np.rint rounds each element of an array to the nearest integer, but it returns the result as a float. If a number is exactly halfway between two integers, it rounds to the nearest even integer.

In [59]:
np.rint(f_array)

array([ 0.,  1.,  3., -3.])

np.fix in NumPy rounds an array of floating-point numbers toward zero, element-wise. It behaves like truncation, where the fractional part of each number is removed without rounding up or down.

In [60]:
np.fix(f_array)

array([ 0.,  1.,  2., -3.])

In [61]:
np.ceil(f_array)

array([ 1.,  2.,  3., -3.])

In [62]:
np.floor(f_array)

array([ 0.,  1.,  2., -4.])

In [63]:
np.sum(f_array)

1.1111000000000004

In [64]:
np.sum([2,3])

5

In [65]:
[2,3,np.nan]

[2, 3, nan]

In [66]:
np.nansum([2,3,np.nan])

5.0

In [67]:
a=np.array([2,3,1.5])
np.exp(a)

array([ 7.3890561 , 20.08553692,  4.48168907])

In [68]:
np.log(a) #log e

array([0.69314718, 1.09861229, 0.40546511])

In [69]:
a=np.array([2,3,1.5,100])
np.log10(a)

array([0.30103   , 0.47712125, 0.17609126, 2.        ])

### Statistical Functions

Min & Max

In [70]:
arr = np.array([[3,7,5],[8,4,3],[2,4,9]]) 
arr

array([[3, 7, 5],
       [8, 4, 3],
       [2, 4, 9]])

In [71]:
print('min: ', np.amin(arr))
print('max: ', np.amax(arr))

min:  2
max:  9


In [72]:
arr

array([[3, 7, 5],
       [8, 4, 3],
       [2, 4, 9]])

In [73]:
# min function along the axis c-0, r-1
print(np.amin(arr, 0))
print(np.amin(arr, 1))

[2 4 3]
[3 3 2]


In [74]:
# max function along the axis c-0, r-1
print(np.amax(arr, 0))
print(np.amax(arr, 1))

[8 7 9]
[7 8 9]


median

In [75]:
arr = np.array([[3,7,5],[8,4,3],[2,4,9]]) 
arr

array([[3, 7, 5],
       [8, 4, 3],
       [2, 4, 9]])

In [76]:
print(np.median(arr))
print(np.median(arr, 0))
print(np.median(arr, 1))

4.0
[3. 4. 5.]
[5. 4. 4.]


mean

In [77]:
arr = np.array([[3,7,5],[8,4,3],[2,4,9]]) 
arr

array([[3, 7, 5],
       [8, 4, 3],
       [2, 4, 9]])

In [78]:
print(np.mean(arr))
print(np.mean(arr, 0))
print(np.mean(arr, 1))

5.0
[4.33333333 5.         5.66666667]
[5. 5. 5.]


In [79]:
arr = np.array([np.nan,221,13,245,0,1])
print(np.mean(arr))
print(np.nanmean(arr))

nan
96.0


average

In [80]:
arr = np.array([1,2,3, 4])
np.average(arr)

2.5

In [81]:
np.average(arr, weights=[4,3,2,1])

2.0

In [82]:
arr = np.array([[3,7,5],[8,4,3],[2,4,9]]) 
arr

array([[3, 7, 5],
       [8, 4, 3],
       [2, 4, 9]])

In [83]:
np.average(arr)

5.0

In [84]:
np.average(arr, axis=0, weights=[4,3,2])

array([4.44444444, 5.33333333, 5.22222222])

In [85]:
np.average(arr, axis=1, weights=[4,3,2])

array([4.77777778, 5.55555556, 4.22222222])

Standard Deviation

In [86]:
arr = np.array([221,13,44,245,0,1])
np.std(arr)

104.25076605100907

Here’s a table summarizing the commonly used NumPy functions, including **syntax** and **examples**:

| **Category**         | **Function**       | **Syntax**                                | **Example**                                                                                           |
|-----------------------|--------------------|-------------------------------------------|-------------------------------------------------------------------------------------------------------|
| **Array Creation**    | `np.array()`       | `np.array(data)`                          | `np.array([1, 2, 3])` → `[1 2 3]`                                                                    |
|                       | `np.zeros()`       | `np.zeros(shape)`                         | `np.zeros((2, 3))` → `[[0. 0. 0.] [0. 0. 0.]]`                                                      |
|                       | `np.ones()`        | `np.ones(shape)`                          | `np.ones((2, 3))` → `[[1. 1. 1.] [1. 1. 1.]]`                                                       |
|                       | `np.empty()`       | `np.empty(shape)`                         | `np.empty((2, 2))` → Random uninitialized values                                                     |
|                       | `np.arange()`      | `np.arange(start, stop, step)`            | `np.arange(0, 5, 1)` → `[0 1 2 3 4]`                                                                |
|                       | `np.linspace()`    | `np.linspace(start, stop, num)`           | `np.linspace(0, 1, 5)` → `[0.   0.25 0.5  0.75 1.  ]`                                               |
|                       | `np.eye()`         | `np.eye(n)`                               | `np.eye(3)` → `[[1. 0. 0.] [0. 1. 0.] [0. 0. 1.]]`                                                  |
| **Array Manipulation**| `np.reshape()`     | `np.reshape(array, new_shape)`            | `np.reshape([1, 2, 3, 4], (2, 2))` → `[[1 2] [3 4]]`                                                |
|                       | `np.transpose()`   | `np.transpose(array)`                     | `np.transpose([[1, 2], [3, 4]])` → `[[1 3] [2 4]]`                                                  |
|                       | `np.flatten()`     | `array.flatten()`                         | `np.array([[1, 2], [3, 4]]).flatten()` → `[1 2 3 4]`                                                |
|                       | `np.concatenate()` | `np.concatenate((arr1, arr2), axis)`      | `np.concatenate(([1, 2], [3, 4]))` → `[1 2 3 4]`                                                    |
|                       | `np.split()`       | `np.split(array, sections)`               | `np.split([1, 2, 3, 4], 2)` → `[array([1, 2]), array([3, 4])]`                                       |
| **Math Operations**   | `np.add()`         | `np.add(arr1, arr2)`                      | `np.add([1, 2], [3, 4])` → `[4 6]`                                                                  |
|                       | `np.subtract()`    | `np.subtract(arr1, arr2)`                 | `np.subtract([5, 7], [3, 4])` → `[2 3]`                                                             |
|                       | `np.multiply()`    | `np.multiply(arr1, arr2)`                 | `np.multiply([1, 2], [3, 4])` → `[3 8]`                                                             |
|                       | `np.divide()`      | `np.divide(arr1, arr2)`                   | `np.divide([6, 8], [3, 4])` → `[2. 2.]`                                                             |
|                       | `np.power()`       | `np.power(base, exp)`                     | `np.power([2, 3], 2)` → `[4 9]`                                                                     |
|                       | `np.sqrt()`        | `np.sqrt(arr)`                            | `np.sqrt([4, 9])` → `[2. 3.]`                                                                       |
|                       | `np.exp()`         | `np.exp(arr)`                             | `np.exp([1, 2])` → `[ 2.71828183  7.3890561 ]`                                                      |
|                       | `np.log()`         | `np.log(arr)`                             | `np.log([1, np.e])` → `[0. 1.]`                                                                     |
| **Statistics**        | `np.mean()`        | `np.mean(arr)`                            | `np.mean([1, 2, 3])` → `2.0`                                                                        |
|                       | `np.median()`      | `np.median(arr)`                          | `np.median([1, 2, 3, 4])` → `2.5`                                                                   |
|                       | `np.std()`         | `np.std(arr)`                             | `np.std([1, 2, 3])` → `0.816496580927726`                                                           |
|                       | `np.var()`         | `np.var(arr)`                             | `np.var([1, 2, 3])` → `0.6666666666666666`                                                          |
|                       | `np.min()`         | `np.min(arr)`                             | `np.min([3, 1, 2])` → `1`                                                                           |
| **Logical Operations**| `np.where()`       | `np.where(condition, x, y)`               | `np.where([True, False], 1, 0)` → `[1 0]`                                                           |
|                       | `np.all()`         | `np.all(arr)`                             | `np.all([True, True])` → `True`                                                                     |
|                       | `np.any()`         | `np.any(arr)`                             | `np.any([False, True])` → `True`                                                                    |
| **Random**            | `np.random.rand()` | `np.random.rand(dims)`                    | `np.random.rand(2, 2)` → Random values in [0, 1)                                                   |
|                       | `np.random.randint()`| `np.random.randint(low, high, size)`      | `np.random.randint(0, 10, 3)` → Random integers like `[3 8 5]`                                       |
|                       | `np.random.choice()`| `np.random.choice(arr, size)`             | `np.random.choice([1, 2, 3], 2)` → Random selection like `[1, 3]`                                   |

---

This table captures syntax and examples for most of the frequently used NumPy functions, making it a quick reference for common operations.